### PR TITLE
[WIP] Infer nullability from NRT

### PR DESCRIPTION
**Pull request by gao-artur**
Replaces #3970.

With this implementation, we can also use a legacy flag to deliver the new feature in v7

### CODE REVIEW

File: src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs [hunk 1]
```diff
@@ -162,15 +162,63 @@ public void auto_register_input_object_graph_type()
 
     [Fact]
     public void accepts_property_expressions()
+    {
+        bool old = GlobalSwitches.InferFieldNullabilityFromNRTAnnotations;
+        GlobalSwitches.InferFieldNullabilityFromNRTAnnotations = false;
```
Comment (gao-artur) [inline src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs:167]: TODO: move to `"StaticTests"`

Comment (Shane32) [reply src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs:167]: lol - i checked already - `ComplexGraphTypeTests` is already marked with `[Collection("StaticTests")]`

**Comment by Shane32** [comment]
> Replaces #3970.
> 
> With this implementation, we can also use a legacy flag to deliver the new feature in v7

Let's just have a PR target `master`, then after it's merged, add a PR to change the default flag (and add migration notes) for v8.

File: src/GraphQL/Types/Composite/ComplexGraphType.cs [hunk 2]
```diff
@@ -561,15 +561,30 @@ public virtual FieldType AddField(FieldType fieldType)
...
     {
         try
         {
-            if (type == null)
-                type = typeof(TProperty).GetGraphTypeFromType(nullable, this is IInputObjectGraphType ? TypeMappingMode.InputType : TypeMappingMode.OutputType);
+            if (type == null && nullable == null && GlobalSwitches.InferFieldNullabilityFromNRTAnnotations)
+            {
+                var typeInfo = AutoRegisteringHelper.GetTypeInformation(((MemberExpression)expression.Body).Member, this is IInputObjectGraphType);
+                type = typeInfo.ConstructGraphType();
+            }
```
Comment (Shane32) [inline src/GraphQL/Types/Composite/ComplexGraphType.cs:582]: I think we need a little more code in here, along these lines -- see example below: (unfinished)

```suggestion
            if (type == null && nullable == null && GlobalSwitches.InferFieldNullabilityFromNRTAnnotations)
            {
                if (expression.Body is MemberExpression memberExpression)
                {
                    var typeInfo = AutoRegisteringHelper.GetTypeInformation(memberExpression.Member, this is IInputObjectGraphType);
                    type = typeInfo.ConstructGraphType();
                }
                else if (typeof(TProperty).IsValueType)
                {
                    nullable = Nullable.GetUnderlyingType(typeof(TProperty)) != null;
                    // call GetGraphTypeFromType
                }
                else // non-member expression such as: x => "Sample"
                {
                    nullable = false; // fallback to legacy behavior
                    // call GetGraphTypeFromType
                }
            }
```

**Comment by Shane32** [comment]
@gao-artur  This PR is closed so I'm deleting the branch.  You can always restore it if need be.  Ditto with other old branches.

**Pull request closed without merging**